src/party.rs

```diff
@@ -24,67 +62,876 @@ pub struct BallotConfig {
         }
+
+        None
+    }
+
+    fn get_value(&self) -> V {
+        self.value_selector.select(&self.parties_voted_before)
     }
 
-    /// Start the party.
-    pub fn start(&mut self) {
-        // TODO: launch party
+    /// Start the next ballot. It's expected from the external system to re-run ballot protocol in
+    /// case of failed ballot.
+    pub async fn launch_ballot(&mut self) -> Result<Option<V>, BallotError> {
+        self.prepare_next_ballot();
+
+        while self.is_launched() {
+            if let Ok(msg_wire) = self.msg_in_receiver.try_recv() {
+                if let Err(err) = self.update_state(msg_wire.content_bytes, msg_wire.routing) {
+                    self.status = PartyStatus::Failed;
+                    return Err(err);
+                }
+            }
+
+            if let Ok(event) = self.event_receiver.try_recv() {
+                if let Err(err) = self.follow_event(event) {
+                    self.status = PartyStatus::Failed;
+                    return Err(err);
+                }
+            }
+
+            // TODO: Emit events to run ballot protocol according to the ballot configuration
+            if self.event_sender.send(PartyEvent::Launch1a).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Launch1a event".into(),
+                ));
+            }
+            if self.event_sender.send(PartyEvent::Launch1b).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Launch1b event".into(),
+                ));
+            }
+            if self.event_sender.send(PartyEvent::Launch2a).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Launch2a event".into(),
+                ));
+            }
+            if self.event_sender.send(PartyEvent::Launch2av).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Launch2av event".into(),
+                ));
+            }
+            if self.event_sender.send(PartyEvent::Launch2b).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Launch2b event".into(),
+                ));
+            }
+            if self.event_sender.send(PartyEvent::Finalize).is_err() {
+                self.status = PartyStatus::Failed;
+                return Err(BallotError::Communication(
+                    "Failed to send Finalize event".into(),
+                ));
+            }
+        }
+
+        Ok(self.get_value_selected())
+    }
+
+    /// Prepare state before running a ballot
+    fn prepare_next_ballot(&mut self) {
```

Comment:
I might call this `reset()` or something to be more clear as to what the function is actually doing.

Reply:
I think `reset` shall imply rollback to a previous state, however in this case we are incrementing the ballot number:

```rust
self.ballot += 1;
```

Thus, maybe it is better to leave it as is, or come up with a shorter name